File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/StopDetailsRoutesView.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.ModalRoutes
import com.mbta.tid.mbta_app.android.MyApplicationTheme
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.PatternsByStop
import com.mbta.tid.mbta_app.model.Prediction
import com.mbta.tid.mbta_app.model.RealtimePatterns
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.StopDetailsDepartures
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.TripDetailsFilter
import com.mbta.tid.mbta_app.model.UpcomingTrip
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import kotlin.time.Duration.Companion.minutes
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant

@Composable
fun StopDetailsRoutesView(
    departures: StopDetailsDepartures?,
    global: GlobalResponse?,
    now: Instant,
    stopFilter: StopDetailsFilter?,
    tripFilter: TripDetailsFilter?,
    pinRoute: (String) -> Unit,
    pinnedRoutes: Set<String>,
    updateStopFilter: (StopDetailsFilter?) -> Unit,
    openAlertDetails: (ModalRoutes.AlertDetails) -> Unit
) {
    if (departures == null) {
        LoadingStopDetailsView(stopFilter, tripFilter)
    } else if (stopFilter != null) {
        StopDetailsFilteredRouteView(
            departures = departures,
            global = global,
            now = now,
            stopFilter = stopFilter,
            tripFilter = tripFilter,
            updateStopFilter,
            openAlertDetails
        )
    } else {
        LazyColumn(Modifier.padding(top = 16.dp).background(colorResource(R.color.fill1))) {
            items(departures.routes, key = { it.routeIdentifier }) { patternsByStop ->
                StopDetailsRouteView(
                    patternsByStop,
                    now,
                    pinned = pinnedRoutes.contains(patternsByStop.routeIdentifier),
                    onPin = pinRoute,
                    updateStopFilter
                )
            }
        }
    }
}

@Preview
@Composable
private fun StopDetailsRoutesViewPreview() {
    val objects = ObjectCollectionBuilder()

    val route1 =
        objects.route {
            color = "00843D"
            longName = "Green Line B"
            textColor = "FFFFFF"
            type = RouteType.LIGHT_RAIL
        }
    val route2 =
        objects.route {
            color = "FFC72C"
            shortName = "57"
            textColor = "000000"
            type = RouteType.BUS
        }
    val stop = objects.stop()
    val trip1 = objects.trip()
    val prediction1 =
        objects.prediction {
            trip = trip1
            departureTime = Clock.System.now() + 5.minutes
        }
    val trip2 = objects.trip()
    val schedule2 =
        objects.schedule {
            trip = trip2
            departureTime = Clock.System.now() + 10.minutes
        }
    val trip3 = objects.trip()
    val prediction2 =
        objects.prediction {
            trip = trip3
            departureTime = Clock.System.now() + 8.minutes
        }
    val trip4 = objects.trip()
    val schedule3 =
        objects.schedule {
            trip = trip4
            departureTime = Clock.System.now() + 10.minutes
        }
    val prediction3 =
        objects.prediction {
            trip = trip4
            departureTime = null
            arrivalTime = null
            scheduleRelationship = Prediction.ScheduleRelationship.Cancelled
        }

    MyApplicationTheme {
        StopDetailsRoutesView(
            departures =
                StopDetailsDepartures(
                    listOf(
                        PatternsByStop(
                            route1,
                            stop,
                            listOf(
                                RealtimePatterns.ByHeadsign(
                                    route1,
                                    "A",
                                    null,
                                    emptyList(),
                                    listOf(UpcomingTrip(trip1, prediction1))
                                )
                            )
                        ),
                        PatternsByStop(
                            route2,
                            stop,
                            listOf(
                                RealtimePatterns.ByHeadsign(
                                    route2,
                                    "B",
                                    null,
                                    emptyList(),
                                    listOf(UpcomingTrip(trip3, prediction2))
                                ),
                                RealtimePatterns.ByHeadsign(
                                    route2,
                                    "C",
                                    null,
                                    emptyList(),
                                    listOf(UpcomingTrip(trip2, schedule2))
                                ),
                                RealtimePatterns.ByHeadsign(
                                    route2,
                                    "D",
                                    null,
                                    emptyList(),
                                    listOf(UpcomingTrip(trip4, schedule3, prediction3))
                                )
                            )
                        )
                    )
                ),
            global = null,
            now = Clock.System.now(),
            stopFilter = null,
            tripFilter = null,
            pinRoute = {},
            pinnedRoutes = emptySet(),
            updateStopFilter = {},
            openAlertDetails = {}
        )
    }
}

File: iosApp/iosApp/Pages/LegacyStopDetails/StopDetailsRoutesView.swift
//
//  StopDetailsRoutesView.swift
//  iosApp
//
//  Created by Horn, Melody on 2024-04-03.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Foundation
import shared
import SwiftUI

struct StopDetailsRoutesView: View {
    let departures: StopDetailsDepartures
    let global: GlobalResponse?
    let now: Instant
    var filter: StopDetailsFilter?
    var setFilter: (StopDetailsFilter?) -> Void
    let pushNavEntry: (SheetNavigationStackEntry) -> Void
    let pinRoute: (String) -> Void
    var pinnedRoutes: Set<String> = []

    var body: some View {
        if let filter {
            StopDetailsFilteredRouteView(
                departures: departures,
                global: global,
                now: now,
                filter: filter,
                setFilter: setFilter,
                pushNavEntry: pushNavEntry,
                pinned: pinnedRoutes.contains(filter.routeId)
            )
        } else {
            ZStack {
                Color.fill1.ignoresSafeArea(.all)
                ScrollView {
                    LazyVStack(spacing: 0) {
                        ForEach(departures.routes, id: \.routeIdentifier) { patternsByStop in
                            StopDetailsRouteView(
                                patternsByStop: patternsByStop,
                                now: now,
                                pushNavEntry: pushNavEntry,
                                pinned: pinnedRoutes.contains(patternsByStop.routeIdentifier),
                                onPin: pinRoute
                            )
                        }
                    }.padding(.top, 16)
                }
            }
        }
    }
}

#Preview {
    let objects = ObjectCollectionBuilder()
    let route1 = objects.route { route in
        route.color = "00843D"
        route.longName = "Green Line B"
        route.textColor = "FFFFFF"
        route.type = .lightRail
    }
    let route2 = objects.route { route in
        route.color = "FFC72C"
        route.shortName = "57"
        route.textColor = "000000"
        route.type = .bus
    }
    let stop = objects.stop { _ in }
    let trip1 = objects.trip { _ in }
    let prediction1 = objects.prediction { prediction in
        prediction.trip = trip1
        prediction.departureTime = (Date.now + 5 * 60).toKotlinInstant()
    }
    let trip2 = objects.trip { _ in }
    let schedule2 = objects.schedule { schedule in
        schedule.trip = trip2
        schedule.departureTime = (Date.now + 10 * 60).toKotlinInstant()
    }
    let trip3 = objects.trip { _ in }
    let prediction2 = objects.prediction { prediction in
        prediction.trip = trip3
        prediction.departureTime = (Date.now + 8 * 60).toKotlinInstant()
    }
    let trip4 = objects.trip { _ in }
    let schedule3 = objects.schedule { schedule in
        schedule.trip = trip4
        schedule.departureTime = (Date.now + 10 * 60).toKotlinInstant()
    }
    let prediction3 = objects.prediction { prediction in
        prediction.trip = trip4
        prediction.departureTime = nil
        prediction.arrivalTime = nil
        prediction.scheduleRelationship = .cancelled
    }

    return StopDetailsRoutesView(departures: .init(routes: [
        .init(route: route1, stop: stop, patterns: [
            .ByHeadsign(
                route: route1,
                headsign: "A",
                line: nil,
                patterns: [],
                upcomingTrips: [.init(trip: trip1, prediction: prediction1)]
            ),
        ], elevatorAlerts: []),
        .init(route: route2, stop: stop, patterns: [
            .ByHeadsign(
                route: route2,
                headsign: "B",
                line: nil,
                patterns: [],
                upcomingTrips: [.init(trip: trip3, prediction: prediction2)]
            ),
            .ByHeadsign(
                route: route2,
                headsign: "C",
                line: nil,
                patterns: [],
                upcomingTrips: [.init(trip: trip2, schedule: schedule2)]
            ),
            .ByHeadsign(
                route: route2,
                headsign: "D",
                line: nil,
                patterns: [],
                upcomingTrips: [
                    .init(trip: trip4, schedule: schedule3, prediction: prediction3),
                ]
            ),
        ], elevatorAlerts: []),
    ]), global: nil, now: Date.now.toKotlinInstant(), filter: nil, setFilter: { _ in }, pushNavEntry: { _ in },
    pinRoute: { routeId in print("Pinned route \(routeId)") }).font(Typography.body)
}
